extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/CompletableFutureTest.java

```diff
@@ -0,0 +1,154 @@
+import org.eclipse.microprofile.graphql.Query;
+import org.eclipse.microprofile.graphql.Source;
+import org.hamcrest.CoreMatchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.filter.log.ResponseLoggingFilter;
+
+/**
+ * Make sure that CompletableFuture works
+ */
+public class CompletableFutureTest extends AbstractGraphQLTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(BookGraphQLApi.class, Book.class)
+                    .addAsResource(new StringAsset(getPropertyAsString()), "application.properties")
+                    .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
+
+    @Test
+    public void testSourcePost() {
+        String booksRequest = getPayload("{\n" +
+                "    books {\n" +
+                "       title\n" +
+                "       published\n" +
+                "       buyLink\n" +
+                "       authors {\n" +
+                "           name\n" +
+                "           bornName\n" +
+                "       }\n" +
+                "    }\n" +
+                "}");
+
+        RestAssured.given()
+                .filter(new ResponseLoggingFilter())
+                .when()
+                .accept(MEDIATYPE_JSON)
+                .contentType(MEDIATYPE_JSON)
+                .body(booksRequest)
+                .post("/graphql")
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body(CoreMatchers.containsString("Lord of the Flies"),
+                        CoreMatchers.containsString("https://www.amazon.com/s?k=Lord+of+the+Flies&i=stripbooks-intl-ship"), // Test source
+                        CoreMatchers.containsString("William Gerald Golding")); // Test batch
+
+    }
+
+    @GraphQLApi
+    public static class BookGraphQLApi {
+
+        @Query
+        public CompletableFuture<List<Book>> getBooks() {
+            return CompletableFuture.supplyAsync(() -> new ArrayList<>(BOOKS.values()));
+        }
+
+        @Query
+        public CompletableFuture<Book> getBook(String name) {
+            return CompletableFuture.supplyAsync(() -> BOOKS.get(name));
+        }
+
+        public CompletableFuture<String> getBuyLink(@Source Book book) {
+            String title = book.title.replaceAll(" ", "+");
+            return CompletableFuture.supplyAsync(() -> String.format(AMAZON_SEARCH_FORMAT, title));
+        }
+
+        public CompletableFuture<List<List<Author>>> getAuthors(@Source List<Book> books) {
```

Comment:
Question: Is this a common construct in GraphQL: List<List<X>> looks like a bi-dimensional array. Shouldn't we have a class for this?

Also for me, it looks like we need streams ability (otherwise you may end up with OOM)

Reply:
This is a batch function, so the first List is because it's a batch. The second one is because a book can have many authors. This comes from graphql-java that we use underneath : https://www.graphql-java.com/documentation/v16/batching/

Reply:
The more I read about graphql-java more I realize that it makes strong assumptions about synchronicity. How coupled is the implementation?

Reply:
At this point, it will be a huge effort to replace graphql-java. There is no other implementation in Java of the graphql spec, and our implementation of the MP Spec is build on graphql-java. So we can write our own implementation of the graphql spec if we want, but that would take some time :)